test/capwap_loc_provider_SUITE.erl

```diff
@@ -0,0 +1,170 @@
+    ok;
+end_per_testcase(_, _) ->
+    meck:expect(test_loc_handler, content, fun() ->  meck:passthrough([]) end),
+    ok.
+
+%% ==============================================
+%% Tests
+%% ==============================================
+
+test_cache() ->
+    [{doc, "Test that the cache is used and refreshed appropriately"}].
+test_cache(_) ->
+    V1 = #{<<"latitude">> => <<"001">>,  <<"longitude">> => <<"002">>},
+    V2 = #{<<"latitude">> => <<"100">>,  <<"longitude">> => <<"200">>},
+    meck:expect(test_loc_handler, content, fun() -> jsx:encode(V1) end),
+    {location, <<"001">>, <<"002">>} = capwap_loc_provider:get_loc(<<"device">>),
+    meck:expect(test_loc_handler, content, fun() -> jsx:encode(V2) end),
+    %% Note we change the value right now to catch errors in caching
+    timer:sleep(1000),
+    %% Cache hasn't expired
+    {location, <<"001">>, <<"002">>} = capwap_loc_provider:get_loc(<<"device">>),
+    timer:sleep(10000),
+    %% Cache has expired now
+    {location, <<"100">>, <<"200">>} = capwap_loc_provider:get_loc(<<"device">>),
+    meck:unload(test_loc_handler),
+    ok.
+
+test_config_chain() ->
+    [{doc, "Test that the next provider is used in case of error"}].
+test_config_chain(_) ->
+    NonConformant = #{<<"error">> => <<"errorMsg">>},
+    meck:expect(test_loc_handler, content, fun() -> jsx:encode(NonConformant) end),
+    {location, <<"123">>, <<"456">>} = capwap_loc_provider:get_loc(<<"device">>).
+
+test_reconfig() ->
+    [{doc, "Test that config can be modified at runtime"}].
+test_reconfig(Config) ->
+    RefreshTime = 5000,
+    LocConfig = #{providers => [
+	{capwap_loc_provider_default, #{default_loc => {location, Lat = <<"123">>, Long = <<"456">>}}}
+      ],
+      refresh => RefreshTime},
+    capwap_loc_provider:load_config(LocConfig),
+    {location, <<"123">>, <<"456">>} = capwap_loc_provider:get_loc(<<"device">>).
+
+test_error() ->
+    [{doc, "Test that an error is returned in case of no valid providers"}].
+test_error(Config) ->
+    RefreshTime = 5000,
+    LocConfig = #{providers => [
+      ],
+      refresh => RefreshTime},
+    capwap_loc_provider:load_config(LocConfig),
+    {error, no_more_providers} = capwap_loc_provider:get_loc(<<"device">>).
+
+test_existing_server_no_cache() ->
+    [{doc, "Test a real endpoint without using a cache"}].
+test_existing_server_no_cache(Config) ->
+    %% The service needs to be exposed through kubectl port-forward
+    RefreshTime = 5000,
+    LocConfig = #{providers => [
+	{capwap_loc_provider_http, #{uri => "http://127.0.0.1:9999", timeout => 30000}}
+      ],
+      refresh => RefreshTime},
+    capwap_loc_provider:load_config(LocConfig),
+    GetAndPrint = fun(Id) ->
+        {location, Lat, Long} = capwap_loc_provider:get_loc(Id, false),
+	ct:pal("Location for ~p retrieved: {~p - ~p}~n", [Id, Lat, Long])
+    end,
+    lists:foreach(GetAndPrint, [<<"00112B0159E5">>, <<"00112B0158EF">>, <<"00112B028E3F">>]).
```

Comment:
please remove artefacts of real world entities.

Reply:
Done, thanks.